rfc.md

```diff
@@ -0,0 +1,215 @@
+# biologica 2.0 RFC
+
+### Motivation 1
+
+To have a more maintainable library with type definitions
+
+### Motivation 2
+
+I have always hated needing to transform organisms back and forth between the
+BioLogica.Organism class objects and plain-object formats. Frequently in Geni*
+we need to turn objects into BioLogica.Organisms before we can use them, or,
+worse, check whether we've already been passed an Organism or a plain object.
+
+For this reason it would be nice to define an Organism so that it was simply a
+typed object, which could be serialized, read, and used without any further
+transformations.
+
+As a result of this, it seems that the library should consist only of pure
+functions which take Organisms (or other typed objects) as inputs (e.g.
+`createGametes(Organism)` instead of `Organism.createGametes()`). A worry is that
+this may result in some awkward syntax.
+
+## Organism definition and creation
+
+### Organism
+```
+  {
+    alleles: "a-a b-b ...",
+    sex: "female"
+    // optional
+    phenotype: {
+      trait: characteristic
+    },
+    // optional
+    secondXAlleles: "c d ..."
+  }
+```
+
+* alleles: string  
+   Fully-defined allele string. Two organisms with the same allele string will always
+   have the same phenotype. **Or should this be called `genotype`**?
+* sex: string  
+* phenotype: { [s: string]: string; }  
+   It's useful to be able to read the phenotype of an organism easily, e.g.
+   `createOrganism(...).phenotype.color` instead of
+   `org = createOrganism(...); getCharacteristic(org, 'wings')`, so it seems to make
+   sense for this to be a part of any organism generated by the library. However, it
+   is redundant information, so it probably doesn't need to be passed in when using
+   an organism.
+* secondXAlleles: string  
+   This formalizes the pattern from geni* that allows us to switch sexes without changing the
+   phenotype. **Maybe this is unneeded?** We could presumably also have a `changeSex` function
+   that explictly picks second-x-alleles that don't change the phenotype. Do we need this for
+   anything else?
```

Comment:
That is generally the direction I was musing along, but I don't think I like that direction after further consideration. The need for the `secondXAlleles` only arises in situations in which we give the user the ability to change sex without changing alleles and want a transition from F => M => F to return to the original organism, i.e. it's a transitory context-dependent requirement. As a thought experiment, in a species for which the X & Y chromosomes contained completely different genes, the transitory requirement would be to store an entire extra chromosome. In the end, I'm inclined to think it shouldn't be part of the genotype, and maybe not even part of the organism, but perhaps would be part of a proto-organism object that contained additional information used in situations where organisms are user-manipulable.

Reply:
Ok, so maybe that gets handled by the outer application, and the only thing the library should support is some easy-to-use functions to make this easy, like maybe `getChromosomeGenotype({chromosome: 'XY', side: 'right'})` when the drake is about to lose those alleles, and `setChromosomeGenotype({chromosome: 'XY', side: 'right', alleles: '...'})` when the drake switches back to female.

Reply:
...also I think it would make sense for any arbitrary properties to be preserved when drakes are created, so the application could support fields like `secondXAlleles` or `name` or whatever in the Organism object itself, and not need yet-another-wrapper.

I was going to suggest these could be any arbitrary fields, but then applications that add their own fields to organisms might get borked when a newer version of biologica wants to use those same field names, so maybe they should be packaged in `metadata`.

e.g.
```
var org = organism({
  authoredAlleles: "a-a",
  metadata: {
    secondXAlleles: "f-f",
    name: "frankie"
  }
}) 
  => 
  org: Organsim = {
    alleles: "a-a b-b c-c ..."
    sex: "female",
    phenotype: {...},
    metadata: {
      secondXAlleles: "f-f",
      name: "frankie"
    }
  }
```